core/server/common/src/main/java/alluxio/master/PrimarySelectorClient.java

```diff
@@ -174,20 +174,26 @@ public void takeLeadership(CuratorFramework client) throws Exception {
    * @return a new {@link CuratorFramework} client to use for leader selection
    */
   private CuratorFramework getNewCuratorClient() {
-    CuratorFramework client = CuratorFrameworkFactory.newClient(mZookeeperAddress,
-        (int) ServerConfiguration.getMs(PropertyKey.ZOOKEEPER_SESSION_TIMEOUT),
-        (int) ServerConfiguration.getMs(PropertyKey.ZOOKEEPER_CONNECTION_TIMEOUT),
-        new ExponentialBackoffRetry(Constants.SECOND_MS, 3));
-    client.start();
+    LOG.info("Creating new zookeeper client for primary selector {}", mZookeeperAddress);
+    CuratorFrameworkFactory.Builder curatorBuilder = CuratorFrameworkFactory.builder();
+    curatorBuilder.connectString(mZookeeperAddress);
+    curatorBuilder.retryPolicy(new ExponentialBackoffRetry(Constants.SECOND_MS, 3));
+    curatorBuilder
+        .sessionTimeoutMs((int) ServerConfiguration.getMs(PropertyKey.ZOOKEEPER_SESSION_TIMEOUT));
+    curatorBuilder.connectionTimeoutMs(
+        (int) ServerConfiguration.getMs(PropertyKey.ZOOKEEPER_CONNECTION_TIMEOUT));
+    // Force compatibility mode to support writing to 3.4.x servers.
+    curatorBuilder.zk34CompatibilityMode(true);
```

Comment:
Should this be configurable?

Reply:
Curator recipe we're using `LeaderElection` is working fine with these limitations. I don't think we'll ever need to switch this off for zk 3.5.x servers since limited functionality is not that essential. 

For our use of zookeeper via curator:

- `zk34CompatibilityMode` causes some previously internal functionality to be invoked via reflection. Rather than going through exposed API in 3.5.x.
- `dontUseContainerParents` just causes to use specific type of `create` calls rather than calling new overload that has an option to create parents recursively.